src/ngx_http_lua_socket_udp.c

```diff
@@ -961,6 +965,139 @@ ngx_http_lua_socket_udp_send(lua_State *L)
 }
 
 
+
+#define stack_diff 1
+static int
+ngx_http_lua_socket_udp_buff_send(lua_State *L)
+{
+    ssize_t                              n;
+    ngx_http_request_t                  *r;
+    ngx_http_lua_socket_udp_upstream_t  *u;
+    ngx_http_lua_loc_conf_t             *llcf;
+    // add new variable for buffer send
+    ngx_str_t                           query;
+    size_t size;
+    size_t head_len;
+    size_t buf_len;
+    u_char *msg_tmp;
+    const char *header_str;
+    const char *buff_str;
+
+
+    if (lua_gettop(L) < 4 ) {
+        return luaL_error(L, "expecting at least 4 arguments (including the object), "
+                          "but got %d", lua_gettop(L));
+    }
+
+    r = ngx_http_lua_get_req(L);
+    if (r == NULL) {
+        return luaL_error(L, "request object not found");
+    }
+
+    luaL_checktype(L, 1, LUA_TTABLE);
+
+    lua_rawgeti(L, 1, SOCKET_CTX_INDEX);
+    u = lua_touserdata(L, -1);
+    lua_pop(L, 1);
+
+    if (u == NULL || u->udp_connection.connection == NULL) {
+        llcf = ngx_http_get_module_loc_conf(r, ngx_http_lua_module);
+
+        if (llcf->log_socket_errors) {
+            ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
+                          "attempt to send data on a closed socket: u:%p, c:%p",
+                          u, u ? u->udp_connection.connection : NULL);
+        }
+
+        lua_pushnil(L);
+        lua_pushliteral(L, "closed");
+        return 2;
+    }
+
+    if (u->request != r) {
+        return luaL_error(L, "bad request");
+    }
+
+    if (u->ft_type) {
+        u->ft_type = 0;
+    }
+
+    if (u->waiting) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "socket busy");
+        return 2;
+    }
+
+    // check and calculate string len
+    int stack_size = lua_gettop(L);
+    size_t start = luaL_checklong(L, -2);
+    size_t end = luaL_checklong(L, -1);
+    buff_str = luaL_checklstring(L, -3, &buf_len);
+    if (start < 1 || end > buf_len || start > end || start > buf_len)
+    {
+        lua_pushnil(L);
+        lua_pushliteral(L, "start or end index invalid");
+        return 2;
+    }
+    size = (end - start + 1);
+    buf_len = size;
+    if (stack_size > 3 + stack_diff)
+    {
+        for (int i = (-1 * stack_size) + stack_diff; i <= -4; i++)
+        {
+            luaL_checklstring(L, i, &head_len);
+            size += head_len;
+        }
+    }
+    // copy msg
```

Comment:
you should use the string.buffer instead of implement this feature.

Reply:
how to use  string.buffer  instead of implement this feature on  lua land ,can you give some example? the copy occurs on the c land,but i need to split the big string on the  lua land ,and also i need to make the header of the udp package on lua land. if i use the origin udpsock:send() method , sub string ( big_msg )  is inevitable ,.usually i can send big string 
 use udp like below，there is no need to make a new string  pass to c land.
```
for idx = 1, totalPackage do
        local header_binary = struct.pack(">i2c1c32i4c1", packageSize + 42, '1', msg_id, totalPackage, zip_type or '0')
        local serial_binary
        if totalPackage == 1 then
            serial_binary = struct.pack(">i4", 0)
        else
            msgEndIndex = msgStartIndex + packageSize - 1
            serial_binary = struct.pack(">i4", idx - 1)
        end
        ok, err = udpsock:sendbuf(header_binary, serial_binary, big_msg, msgStartIndex, msgEndIndex)
        msgStartIndex = msgEndIndex + 1
    end
```

Reply:
https://luajit.org/ext_buffer.html

Reply:
gotcha